sycl/include/sycl/stl_wrappers/__sycl_cmath_wrapper_impl.hpp

```diff
@@ -0,0 +1,541 @@
+//==------------- __sycl_cmath_wrapper_impl.hpp ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CMATH_FALLBACK_H__
+#define __CMATH_FALLBACK_H__
+
+// This header defines device-side overloads of <cmath> functions based on
+// their equivalent __spirv_ built-ins.
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+// The 'sycl_device_only' attribute enables device-side overloading.
+#define __DPCPP_SYCL_DEVICE __attribute__((sycl_device_only, always_inline))
+#define __DPCPP_SYCL_DEVICE_C                                                  \
+  extern "C" __attribute__((sycl_device_only, always_inline))
+
+// Promotion templates: the C++ standard library provides overloads that allow
+// arguments of math functions to be promoted. Any floating-point argument is
+// allowed to accept any integer type, which should then be promoted to double.
+// When multiple floating point arguments are available passing arguments with
+// different precision should promote to the larger type. The template helpers
+// below provide the machinery to define these promoting overloads.
+template <typename T, bool = (std::is_integral<T>::value ||
+                              std::is_floating_point<T>::value)>
+struct __dpcpp_promote {
```

Comment:
@bader , I think you objected to `dpcpp` somewhere else recently. Can you confirm it's fine here?

Reply:
Yes. Updated identifiers in https://github.com/intel/llvm/pull/18706/commits/ef1e40633a42c161e0dabf5bf92ef1b62e9687a1.